KIPs/kip-37.md

```diff
@@ -0,0 +1,908 @@
+  - An implementation MAY emit the `URI` event during a mint operation but it is NOT mandatory. An observer MAY fetch the metadata uri at mint time from the `uri` function if it was not emitted.
+- The `uri` function SHOULD be used to retrieve values if no event was emitted.
+- The `uri` function MUST return the same value as the latest event for an `_id` if it was emitted.
+- The `uri` function MUST NOT be used to check for the existence of a token as it is possible for an implementation to return a valid string even if the token does not exist.
+
+```solidity
+pragma solidity 0.5.6;
+
+/**
+    Note: The KIP-13 identifier for this interface is 0x0e89341c.
+*/
+interface KIP37Metadata_URI {
+    /**
+        @notice A distinct Uniform Resource Identifier (URI) for a given token.
+        @dev URIs are defined in RFC 3986.
+        The URI MUST point to a JSON file that conforms to the "KIP-37 Metadata URI JSON Schema".
+        @return URI string
+    */
+    function uri(uint256 _id) external view returns (string memory);
+}
+```
+
+#### KIP-37 Metadata URI JSON Schema
+
+This JSON schema is loosely based on the "KIP-17 Metadata JSON Schema", but includes optional formatting to allow for ID substitution by clients. If the string `{id}` exists in any JSON value, it MUST be replaced with the actual token ID, by all client software that follows this standard.
+
+- The string format of the substituted hexadecimal ID MUST be lowercase alphanumeric: `[0-9a-f]` with no 0x prefix.
+- The string format of the substituted hexadecimal ID MUST be leading zero padded to 64 hex characters length if necessary.
+
+```json
+{
+  "title": "Token Metadata",
+  "type": "object",
+  "properties": {
+    "name": {
+      "type": "string",
+      "description": "Identifies the asset to which this token represents"
+    },
+    "decimals": {
+      "type": "integer",
+      "description": "The number of decimal places that the token amount should display - e.g. 18, means to divide the token amount by 1000000000000000000 to get its user representation."
+    },
+    "description": {
+      "type": "string",
+      "description": "Describes the asset to which this token represents"
+    },
+    "image": {
+      "type": "string",
+      "description": "A URI pointing to a resource with mime type image/* representing the asset to which this token represents. Consider making any images at a width between 320 and 1080 pixels and aspect ratio between 1.91:1 and 4:5 inclusive."
+    },
+    "properties": {
+      "type": "object",
+      "description": "Arbitrary properties. Values may be strings, numbers, object or arrays."
+    }
+  }
+}
+```
+
+An example of an KIP-37 Metadata JSON file follows. The properties array proposes some SUGGESTED formatting for token-specific display properties and metadata.
+
+```json
+{
+  "name": "Asset Name",
+  "description": "Lorem ipsum...",
+  "image": "https://s3.amazonaws.com/your-bucket/images/{id}.png",
```

Comment:
@kaikookim There is no field named "decimals" while it is defined in "KIP-37 Metadata URI JSON Schema" above.

Reply:
@kjhman21 Decimals are defined only in metadata, not in contract.

Reply:
I added "decimals" to metadata example.